module5/lessons/Week1/RefactoringPatterns.md

```diff
@@ -0,0 +1,415 @@
+            if (item.Category == "Electronics")
+            {
+                totalPrice += 10;
+            }
+            else if (item.Category == "Clothing")
+            {
+                totalPrice += 5;
+            }
+        }
+
+        // Apply bulk discount for high total prices
+        if (totalPrice > 200)
+        {
+            totalPrice -= totalPrice * 0.1M; // 10% discount
+        }
+
+        // Apply tax
+        totalPrice += totalPrice * 0.08M; // 8% tax
+
+        return totalPrice;
+    }
+}
+
+public class Item
+{
+    public string Name { get; set; }
+    public string Category { get; set; }
+    public decimal Price { get; set; }
+    public int Quantity { get; set; }
+}
+
+class Program
+{
+    static void Main(string[] args)
+    {
+        List<Item> items = new List<Item>
+        {
+            new Item { Name = "Smartphone", Category = "Electronics", Price = 1000, Quantity = 1 },
+            new Item { Name = "T-shirt", Category = "Clothing", Price = 20, Quantity = 3 },
+            // ... add more items ...
+        };
+
+        ShoppingCart cart = new ShoppingCart();
+        decimal total = cart.CalculateTotalPrice();
+
+        Console.WriteLine($"Total Price: {total:C}");
+    }
+}
+
+```
+
+In this example, we have a method `CalculateTotalPrice` that is doing too much.  It is calculating total price **AND** processing additional fees **AND** applying bulk discounts **AND** processing tax.  All those `AND`s indicate that we can extract some functionality from this method.
+
+```c#
+public class ShoppingCart
+{
+    private List<Item> items = new List<Item>();
+
+    public decimal CalculateTotalPrice()
+    {
+        decimal totalPrice = CalculateItemsPriceWithFees();
+
+        totalPrice -= ApplyBulkDiscount(totalPrice);
+
+        totalPrice += ApplyTax(totalPrice);
+
+        return totalPrice;
+    }
+
+    private decimal CalculateItemsPriceWithFees()
+    {
+        decimal totalPrice = 0;
+
+        foreach (var item in items)
+        {
+            totalPrice += item.Price * item.Quantity;
+
+            totalPrice += ApplySpecialFees(item.Category);
+        }
+
+        return totalPrice;
+    }
+
+    private decimal ApplySpecialFees(string category)
+    {
+        if (category == "Electronics")
+        {
+            return 10;
+        }
+        else if (category == "Clothing")
+        {
+            return 5;
+        }
+        return 0;
+    }
+
+    private decimal ApplyBulkDiscount(decimal totalPrice)
+    {
+        if (totalPrice > 200)
+        {
+            return totalPrice * 0.1M; // 10% discount
+        }
+        return 0;
+    }
+
+    private decimal ApplyTax(decimal totalPrice)
+    {
+        return totalPrice * 0.08M; // 8% tax
+    }
+}
+```
+
+With the **extract method** refactoring pattern, our executing code can remain the same, but we have made our classes and methods _much_ easier to read.
+
+## Extract Interfaces
```

Comment:
Reading through this pattern again, it feels pretty nuanced and might be more than our students need. If they are using an interface at all in OOP that feels like a big win! What about switching to something like maybe this one? https://refactoring.guru/decompose-conditional

Reply:
But also it does show a concrete example of some of the SOLID best practices 🤷‍♀️ idk.